Basic GCE PodSecurityPolicy Config

tallclair opened the pull request:
**What this PR does / why we need it**:

This PR lays the foundation for enabling PodSecurityPolicy in GCE and other default deployments. The 3 commits are:

1. Add policies, roles & bindings for the default addons on GCE.
2. Enable the PSP admission controller & load the addon policies when the`ENABLE_POD_SECURITY_POLICY=true` environment variable is set.
3. Support the PodSecurityPolicy in the E2E environment & add PSP tests.

NOTES:

- ~~Depends on https://github.com/kubernetes/kubernetes/pull/52301 for privileged capabilities~~
- ~~Depends on https://github.com/kubernetes/kubernetes/pull/52849 for sane mutations~~
- ~~Depends on https://github.com/kubernetes/kubernetes/pull/53479 for aggregator tests to pass~~
- ~~Depends on https://github.com/kubernetes/kubernetes/pull/54175 for dedicated fluentd service~~ account
- This PR is a fork of https://github.com/kubernetes/kubernetes/pull/46064, credit to @Q-Lee

**Which issue this PR fixes**: #43538

**Release note**:
```release-note
Add support for PodSecurityPolicy on GCE: `ENABLE_POD_SECURITY_POLICY=true` enables the admission controller, and installs policies for default addons.
```

liggitt commented:
Two issues we have to fix before enabling PSP:
* define order when multiple PSPs are usable by the creating user
* split into validation and defaulting passes, and only validate on update

Both would be addressed by https://github.com/kubernetes/kubernetes/pull/50169. I'll try to get that completed in a minimal fashion shortly for discussion.

liggitt commented on cluster/addons/podsecuritypolicies/unprivileged.yaml:
```diff
@@ -0,0 +1,53 @@
+spec:
+  privileged: false
+  allowPrivilegeEscalation: true
+  volumes: # All volume types except HostPath
```
allowing direct use of all these inline volumes is not very unprivileged, is it? don't many of these let you reference arbitrary paths or mount namespaces? cc @kubernetes/sig-storage-pr-reviews

msau42 replied:
Only hostpath volumes let you specify arbitrary paths.  Mount propagation is separately controlled via SecurityContext.Privileged.

liggitt replied:
I meant things like nfs.path, quobyte.volume, vsphereVolume.volumePath, etc, that let a pod author target arbitrary volumes as the node

msau42 replied:
Ah, I see what you mean.  But will restricting the volume types here cause users to break on upgrade?

destijl replied:
So I think the intent of this policy is to document an unprivileged policy for others but also apply it to some of our own components as above. I assume this list is here to allow others to apply the policy and have it work in most cases. Perhaps this is the wrong approach. Would it be more useful to provide the minimal PSP that would let a vanilla K8s cluster function and have users add the extra storage types and other things they need in a new policy? 

As is, it might be hard for a non-expert who wants to lock down their config to reason about what's required for core k8s and what can be safely removed if they aren't using it in their applications.

If we did that would we also need an unprivileged-gce policy for kube-dns etc. where you are currently applying this one?

liggitt replied:
> will restricting the volume types here cause users to break on upgrade?

It is certainly possible. Tightening restrictions on a cluster without any knowledge of what workloads are running while guaranteeing no disruption is impossible. I would expect a rollout like this:
* upgraded clusters do not enable PSP by default
* upgraded clusters can enable PSP with opt-in by users
* new clusters can enable PSP by default

> Would it be more useful to provide the minimal PSP that would let a vanilla K8s cluster function and have users add the extra storage types and other things they need in a new policy?

Absolutely. It is extremely difficult to tighten default policies. I would expect something like:
* a privileged PSP that allows anything, bound to the superuser (and possibly to kubelets, since they can only create mirror pods and should be able to create a mirror pod with any combination of permissions)
* a restricted PSP that only allows "safe" inline volume types (secret, configmap, emptydir, projected, pvc, downwardAPI), no host access, no root permissions, no escalation, etc
* PSPs defined for specific use cases granted to specific service accounts for specific addons.

liggitt commented on cluster/addons/podsecuritypolicies/unprivileged.yaml:
```diff
@@ -0,0 +1,53 @@
+  hostIPC: false
+  hostPID: false
+  runAsUser:
+    rule: 'RunAsAny'
```
I expected RunAsNonRoot

liggitt commented on cluster/addons/podsecuritypolicies/unprivileged.yaml:
```diff
@@ -0,0 +1,53 @@
+    addonmanager.kubernetes.io/mode: Reconcile
+spec:
+  privileged: false
+  allowPrivilegeEscalation: true
```
?

destijl replied:
+1 this seems like a mistake?

tallclair replied:
`allowPrivilegeEscalation: false` doesn't do anything if the container is running as root.

liggitt commented on cluster/gce/config-default.sh:
```diff
@@ -245,8 +245,17 @@ if [[ -n "${GCE_ALPHA_FEATURES:-}" ]]; then
-ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,Priority,ResourceQuota
+ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,Priority
+
+# Optional: use a default, fairly open PodSecurityPolicy
```
what is the intent behind enabling this with an open policy? GCE will have an enormously difficult time tightening it if it does it this way.

tallclair replied:
This comment was out of date.

tallclair replied:
This comment was out of date.

tallclair commented:
I made the suggested changes. Notably, there are now several default PodSecurityPolicies:

- `privileged` - The allow everything policy, used by a couple privileged addons.
- `restricted` - Mostly for example purposes, demonstrates best-practices for hardening containers
- `system.unprivileged-addon` - Generic unprivileged policy for several addons
- per-addon policies - For non-privileged addons that require hostPath volumes or other elevated privileges

tallclair commented:
TODO: Add `metadata-proxy` binding.

@ihmccreery - What service account does the metadata-proxy use?

php-coder commented on cluster/addons/podsecuritypolicies/privileged.yaml:
```diff
@@ -1,23 +1,24 @@
-  allowedCapabilities: null
+  allowPrivilegeEscalation: true
+  allowedCapabilities:
+  - '*'
```
This won't work until we merge https://github.com/kubernetes/kubernetes/pull/51337 Probably this PR should depends on it.

tallclair replied:
Thanks for the heads up. I'm not sure why I thought '*' would work here...

php-coder commented on cluster/addons/podsecuritypolicies/privileged.yaml:
```diff
@@ -1,23 +1,24 @@
+  annotations:
+    kubernetes.io/description: 'privileged allows full unrestricted access to
+      pod features, as if the PodSecurityPolicy controller was not enabled.'
+    seccomp.security.alpha.kubernetes.io/allowedProfileNames: '*'
```
I feel that we miss `apparmor.security.beta.kubernetes.io/allowedProfileNames` here.

tallclair replied:
Unlike with seccomp, the default for AppArmor is allow-all. Going forward, that needs to be the default value to avoid breaking existing profiles.

php-coder commented on cluster/addons/podsecuritypolicies/restricted.yaml:
```diff
@@ -0,0 +1,37 @@
+    addonmanager.kubernetes.io/mode: Reconcile
+spec:
+  privileged: false
+  allowPrivilegeEscalation: false
```
Do we need to also specify `DefaultAllowPrivilegeEscalation: false` here?

tallclair replied:
No, it's defaulted from `allowPrivilegeEscalation` here: https://github.com/kubernetes/kubernetes/blob/98ed5dd8a26929245e4aa30f110242ac7bb7cbbd/pkg/security/podsecuritypolicy/provider.go#L192

tallclair commented:
FYI, added metadata-proxy config. /cc @ihmccreery

dims commented:
/retest

tallclair commented:
Friendly ping. I'm still hoping to get this into 1.8.

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/kube-system-binding.yaml:
```diff
@@ -0,0 +1,17 @@
+subjects:
+- kind: Group
+  # All service accounts in the kube-system namespace are allowed to use this.
+  name: system:authenticated
```
this is more broad than the comment.. intentional?

tallclair replied:
Ah, I didn't know about the serviceaccounts group. Done.

I think this is effectively equivalent though, unless it's possible to run a pod without a service account?

liggitt commented on cluster/addons/podsecuritypolicies/restricted.yaml:
```diff
@@ -0,0 +1,37 @@
+    # add the custom profiles here.
+    seccomp.security.alpha.kubernetes.io/allowedProfileNames: 'docker/default'
+    seccomp.security.alpha.kubernetes.io/defaultProfileName:  'docker/default'
+    apparmor.security.beta.kubernetes.io/defaultProfileName:  'runtime/default'
```
remember that order of PSP's isn't well defined when multiple match yet...  that becomes an issue when some apply defaulting

tallclair replied:
Ack. Added a warning. Since this is "documentation", I think it's OK here.

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/event-exporter-binding.yaml:
```diff
@@ -0,0 +1,16 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: system:podsecuritypolicy:event-exporter
```
logic behind sometimes using `system:podsecuritypolicy:...` as the role name and sometimes just `podsecuritypolicy:...`?

tallclair replied:
The idea was to distinguish those policies that have a very specific use to support system pods from those that are more generally useful. Specifically, the privileged & restricted policies are intended for use outside the kube-system namespace.

Any suggestions for how to make this cleaner?

liggitt replied:
typically, we've tried to namespace roles/bindings to the creating component so that there's no conflict with user-created roles (`system:...` for things created by the apiserver, `gce:...` for things created by gce deployment scripts, `kubeadm:...` for things created by kubeadm, etc)

tallclair replied:
Done. Long-term we should be using namespaces, not prefixes for this (non-namespaced objects might require a different approach -- namespaces in name only?). For now I went with the `gce:` prefix. We may also want to consider fully-qualified names (I.e. k8s.io, cloud.google.com, etc.) at some point.

liggitt commented on cluster/addons/podsecuritypolicies/unprivileged-addon.yaml:
```diff
@@ -0,0 +1,38 @@
+  hostNetwork: false
+  hostIPC: false
+  hostPID: false
+  # TODO: The addons using this profile should not run as root.
```
this TODO will be hard to fix... do we know which addons wouldn't work if you set RunAsNonRoot?

tallclair replied:
I have a google-internal spreadsheet tracking this, I can work on a public version. My guess is this will be pretty straightforward for most non-privileged pods. There are a couple pods that just need root to write to a host-path logfile that's owned by root, but that should be easy to fix.

I wish Kubernetes managed user IDs as a system resource to try and avoid overlapping users. My approach so far has just been to assign everything non-root to `65534` (which is typically `nobody`).

liggitt replied:
> There are a couple pods that just need root to write to a host-path logfile that's owned by root, but that should be easy to fix.

this doesn't allow hostPath volumes... anything else blocking using MustRunAsNonRoot?

tallclair reviewed (commented):
Thanks @liggitt !

dims commented:
/retest

php-coder commented on cluster/addons/podsecuritypolicies/restricted.yaml:
```diff
@@ -0,0 +1,41 @@
+  name: restricted-v1
+  annotations:
+    kubernetes.io/description: 'This policy demonstrates best practices for
+      running Pods with minimal security priviliges.'
```
s/priviliges/privileges/

tallclair replied:
Thanks, done.

php-coder commented on cluster/addons/podsecuritypolicies/restricted.yaml:
```diff
@@ -0,0 +1,41 @@
+metadata:
+  name: restricted-v1
+  annotations:
+    kubernetes.io/description: 'This policy demonstrates best practices for
```
Consider adding `requiredDropCapabilities` to `restricted-v1` PSP in order to conform to best practices:
```yaml
requiredDropCapabilities:
- KILL
- MKNOD
- SETUID
- SETGID
```

tallclair replied:
Hmm, good suggestion, except there doesn't appear to be any way to default the dropped capabilities list. I'm not an advocate of the PSP defaults, but specifying this as is in the restricted profile would require every pod to explicitly list these 4 dropped capabilities....

WDYT?

tallclair replied:
Oops, nvm. I realized `requiredDropCapabilities` is mutating.

In that case, I dropped a lot more, mostly following this guide: http://rhelblog.redhat.com/2016/10/17/secure-your-containers-with-this-one-weird-trick/

Open for discussion, but if there is too much back-and-forth, I might save this for a follow up PR.

php-coder replied:
> Open for discussion, but if there is too much back-and-forth, I might save this for a follow up PR.

Yes, it's better to discuss it separately. In OpenShift we're dropping only these 4 caps (`KILL`, `MKNOD`, `SETUID`, `SETGID`) in the `restricted` SCC.

mikedanese commented:
/approve

deads2k commented:
Continuing the discussion from sig-auth and slack with @tallclair.  This pull is really late for 1.8.0 and the behavioral problems around defaulting and updates are a severe problem.  The inclusion of the sample/bootstrapped PSPs exacerbates the concerns, since it makes it more likely that someone will get bit unexpectedly.  If the resources were not included, I think this pull becomes simply a way to optionally turn on PSP enforcement in GCE and puts that platform on even footing with other deployments.  I still think its too late for 1.8.0, but it seems pretty reasonable (without stock resources) for a 1.8.1.

@tallclair do you have a more complete proposal for PSP bootstrapping (similar to RBAC bootstrapping) across all platforms in 1.9?  I suspect that @liggitt would have ideas in that area too.

dims commented:
@deads2k @mikedanese @tallclair @liggitt - So we can move this out of 1.8 please?

liggitt commented:
> @deads2k @mikedanese @tallclair @liggitt - So we can move this out of 1.8 please?

moved out. can discuss a pure enablement for 1.8.x, and more full-featured enablement in combination with ordering/priority fixes for 1.9

tallclair commented:
SG, I agree this is to late for 1.8.0. However, as the PSP is currently beta, I think we have some responsibility to fix the major issues. I'd like to get https://github.com/kubernetes/kubernetes/pull/52849 (or some variation) cherry-picked into 1.8.x, and then reopen this discussion.

mikedanese commented:
@tallclair I'd also recommend we start testing PSP enablement in test on release-1.8 branch.

tallclair commented:
Enabling PSP in the e2e tests requires a lot more changes than what is here. In the interest of keeping the PR size smaller, I was hoping to merge something like this first, and then following up with a PR to enable it in testing.

tallclair reviewed (commented):
Can we reopen this discussion in the context of 1.9?

destijl commented on cluster/addons/podsecuritypolicies/restricted.yaml:
```diff
@@ -0,0 +1,41 @@
+    # add the custom profiles here.
+    seccomp.security.alpha.kubernetes.io/allowedProfileNames: 'docker/default'
+    apparmor.security.beta.kubernetes.io/allowedProfileNames: 'runtime/default'
+    # WARNING: PodSecurityPolicy ordering is not (yet) well defined, so if
```
link to bug or PR?

liggitt replied:
resolved, can remove this comment?

destijl reviewed (commented):
/lgtm

tallclair commented:
/hold

Wait for https://github.com/kubernetes/kubernetes/pull/52849

Thanks Greg!

tallclair commented:
Added missing `persistent-volume-binder` policies (https://github.com/kubernetes/kubernetes/pull/52367/commits/8baafaff417a624a00a959777c7f7ce6bb1fbbba)

tallclair commented:
@liggitt @destijl @php-coder - PTAL. I've added a couple commits since this was last reviewed: added kube-proxy & node bindings, reduced capabilities in the restricted profile, and (most significantly) E2E tests. Let me know if you'd prefer to see the e2e tests in a separate PR.

I will keep the `do-not-merge/hold` on this until https://github.com/kubernetes/kubernetes/pull/52849 is merged.

tallclair commented:
rebased.

liggitt commented on cluster/gce/config-default.sh:
```diff
@@ -251,8 +251,17 @@ if [[ -n "${GCE_ALPHA_FEATURES:-}" ]]; then
 # Admission Controllers to invoke prior to persisting objects in cluster
-# If we included ResourceQuota, we should keep it at the end of the list to prevent incrementing quota usage prematurely.
-ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,Priority,ResourceQuota
+ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,Priority
```
define ENABLE_POD_SECURITY_POLICY first, then group the ADMISSION_CONTROL building together

tallclair replied:
Removed.

liggitt commented on cluster/gce/config-test.sh:
```diff
@@ -287,8 +287,19 @@ if [[ -n "${GCE_ALPHA_FEATURES:-}" ]]; then
-# If we included ResourceQuota, we should keep it at the end of the list to prevent incrementing quota usage prematurely.
-ADMISSION_CONTROL="${KUBE_ADMISSION_CONTROL:-Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,PodPreset,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,Priority,ResourceQuota}"
+# Optional: enable the PodSecurityPolicy admission contrller.
+ENABLE_POD_SECURITY_POLICY="${ENABLE_POD_SECURITY_POLICY:-false}"
```
only use falls back to empty string, no need to assign to ENABLE_POD_SECURITY_POLICY here

tallclair replied:
Done.

destijl commented:
/lgtm

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/event-exporter-binding.yaml:
```diff
@@ -0,0 +1,16 @@
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
`rbac.authorization.k8s.io/v1` everywhere?

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/fluentd-gcp-binding.yaml:
```diff
@@ -0,0 +1,17 @@
+  name: system:podsecuritypolicy:fluentd-gcp
+subjects:
+- kind: ServiceAccount
+  # TODO: fluentd-gcp should run with a custom service account.
```
I prefer we not start by granting elevated permissions to the default service account... is it small to declare the custom service account?

tallclair replied:
I hear you, but I'm trying to avoid any side affects to this PR. Since this is limited to the kube-system namespace, can I address this in an (immediate) follow up?

liggitt replied:
is it just this?
```diff
diff --git a/cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml b/cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml
index c9cc6bd558..8a0d01df94 100644
--- a/cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml
+++ b/cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml
@@ -1,3 +1,13 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: fluentd-gcp
+  namespace: kube-system
+  labels:
+    k8s-app: fluentd-gcp
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
+---
 apiVersion: extensions/v1beta1
 kind: DaemonSet
 metadata:
@@ -25,6 +35,7 @@ spec:
     spec:
       dnsPolicy: Default
       hostNetwork: true
+      serviceAccountName: fluentd-gcp
       containers:
       - name: fluentd-gcp
         image: gcr.io/google-containers/fluentd-gcp:2.0.9
```

if so, can we open that PR separately and get it merged in parallel with this? seems like it could be in immediately

tallclair replied:
https://github.com/kubernetes/kubernetes/pull/54175

liggitt commented on cluster/addons/podsecuritypolicies/fluentd-gcp.yaml:
```diff
@@ -0,0 +1,39 @@
+    - pathPrefix: /var/log
+    - pathPrefix: /var/lib/docker/containers
+    - pathPrefix: /usr/lib64
+    - pathPrefix: /etc/ssl/certs
```
until we add ability to require readOnly here, this gives write access to the system trust roots, right?

tallclair replied:
Right :frowning_face: I don't think there's any reason fluentd can't just use the certs built into it's container. I'll do that in the follow up mentioned above.

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/node-binding.yaml:
```diff
@@ -0,0 +1,22 @@
+  name: system:podsecuritypolicy:nodes
+  namespace: kube-system
+  annotations:
+    kubernetes.io/description: 'Allow nodes to create privileged mirror pods.'
```
> Allow nodes to create privileged pods. Should be used in combination with the NodeRestriction admission plugin to limit nodes to mirror pods bound to themselves.

tallclair replied:
Done.

liggitt commented on test/e2e/framework/psp_util.go:
```diff
@@ -0,0 +1,71 @@
+
+func IsPodSecurityPolicyEnabled(f *Framework) bool {
+	isPSPEnabledOnce.Do(func() {
+		psp, err := f.ClientSet.ExtensionsV1beta1().PodSecurityPolicies().
```
this makes the name of PSP act like an API, which isn't great. can the e2e test do `[Feature:PodSecurityPolicy]` instead?

tallclair replied:
Then it won't run in the default test suite, which is what I was trying to avoid by doing it this way. The other alternative would be to pipe something through the `TestContext`. WDYT?

liggitt replied:
hmm. If we want to detect enablement, presence of *any* PSP objects seems more reasonable to know if the feature is in use, then have the PSP-specific test create and clean up PSP objects specific to its test rather than depend on deployment-specific ones with well-known names.

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,197 @@
+
+	It("should enforce the restricted PodSecurityPolicy", func() {
+		By("Binding the restricted policy for the test service account")
+		framework.BindClusterRole(f.ClientSet.RbacV1beta1(), "podsecuritypolicy:restricted:v1", ns,
```
I'd expect the e2e test to define the PSPs, roles, and bindings, rather than assuming a particular deployment method or addons

liggitt commented on test/e2e/framework/framework.go:
```diff
@@ -373,6 +374,11 @@ func (f *Framework) CreateNamespace(baseName string, labels map[string]string) (
 	}
+
+	if !f.SkipDefaultPSPBinding {
+		CreateDefaultPSPBinding(f, ns.Name)
```
granting superuser permissions inside every e2e namespace wasn't exactly what I expected. is this an interim step? if so, add a TODO with the plan for scoping this down the tests that require elevated permissions.

@deads2k any feedback based on the process you went through to enable RBAC in e2e?

liggitt replied:
So to summarize
1. enabling PSP in the test env would ensure the addons configured at startup worked with the OOB config. That's good. 
2. The default e2e behavior would grant privileged PSP access to any e2e test namespace that didn't opt out
    - this depends on a well-known or e2e-created privileged PSP… I'd lean toward e2e-created to remove dependencies on cluster config
    - That seems reasonable, at least as a first step, though I'd be interested in a follow up that made e2e tests opt in to elevated pod permissions (similar to how they have to opt into RBAC permissions for pods that make API calls). That will help uncover more gaps, usability issues, and assumptions made in our test images
3. The specific PSP e2e test would opt out of privileged access and exercise PSP limitations. 
    - this depends on the cluster config not giving access to privileged PSP use in arbitrary namespaces by default, which is a good thing to expect

tallclair reviewed (commented):
Comments addressed. PTAL.

Thanks!

tallclair commented:
Removed the restricted profile as it's not actually used, and I don't want it to slow down the rest of the review. I made add it back in a follow up PR, or just turn it into documentation.

tallclair commented:
Rebased. Friendly ping @liggitt

tallclair commented:
https://github.com/kubernetes/kubernetes/pull/53479 merged, which should fix the aggregator failures.

/retest

tallclair commented:
Added dependency on https://github.com/kubernetes/kubernetes/pull/54175

tallclair commented:
All dependencies merged, removed hold.

/retest

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/kube-proxy-binding.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: system:podsecuritypolicy:kube-proxy
```
gce prefix?

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/kube-proxy-binding.yaml:
```diff
@@ -0,0 +1,15 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: podsecuritypolicy:privileged
```
gce prefix?

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/kube-system-binding.yaml:
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: system:podsecuritypolicy:unprivileged-addon
```
gce prefix?

liggitt commented on cluster/addons/podsecuritypolicies-rbac/gce/kube-system-binding.yaml:
```diff
@@ -0,0 +1,17 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: system:podsecuritypolicy:unprivileged-addon
```
gce prefix?

liggitt commented on cluster/addons/podsecuritypolicies/persistent-volume-binder-role.yaml:
```diff
@@ -0,0 +1,20 @@
+# namespace, so this is a ClusterRole.
+kind: ClusterRole
+metadata:
+  name: system:podsecuritypolicy:persistent-volume-binder
```
typically don't use system prefixes for things created outside the API server / controller manager

liggitt commented on cluster/addons/podsecuritypolicies/persistent-volume-binder-role.yaml:
```diff
@@ -0,0 +1,20 @@
+- apiGroups:
+  - extensions
+  resourceNames:
+  - system.persistent-volume-binder
```
same here

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,333 @@
+	BeforeEach(func() {
+		if !framework.IsPodSecurityPolicyEnabled(f) {
+			framework.Skipf("PodSecurityPolicy not enabled")
+		}
```
these tests assume control over PSP permissions using RBAC... e2e today doesn't actually require RBAC. might need a similar skip-unless-RBAC test in the future... not sure

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,333 @@
+			UserName: serviceaccount.MakeUsername(ns, "default"),
+			Groups:   serviceaccount.MakeGroupNames(ns),
+		}
+		c, err = clientset.NewForConfig(config)
```
I don't know ginkgo well enough to know if sharing state between BeforeEach closures (possibly invoked multiple times?) and `It("...", func)` blocks is ok... it certainly smells funny

liggitt replied:
for framework, sure, but you're assigning to `c`... run in parallel, still feels like it could cause races

tallclair replied:
From the [ginkgo docs](https://onsi.github.io/ginkgo/):

> - Ginkgo makes extensive use of closures to allow you to build descriptive test suites.
> - You should make use of Describe and Context containers to expressively organize the behavior of your code.
> - You can use BeforeEach to set up state for your specs. You use It to specify a single spec.
> - In order to share state between a BeforeEach and an It you use closure variables, typically defined at the top of the most relevant Describe or Context container.

Also:
> It is also a mistake to initialize a closure variable in a container block. If one of your Its mutates that variable, subsequent Its will receive the mutated value. This is a case of test pollution and can be hard to track down. **Always initialize your variables in BeforeEach blocks.**

Which makes me think that `It` blocks in a single context are always run serially? Ah, this explains it:

> Ginkgo has support for running specs in parallel. It does this by spawning separate go test processes and serving specs to each process off of a shared queue. This is important for a BDD test framework, as the shared context of the closures does not parallelize well in-process.

So all tests within a single process run serially.

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,333 @@
+
+	It("should allow pods under the privileged PodSecurityPolicy", func() {
+		By("Creating & Binding a privileged policy for the test service account")
+		createAndBindPSP(f, permissivePSPTemplate)
```
create and bind the restricted one as well, and ensure the admitted pod was admitted under the permissive PSP

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,333 @@
+
+func createAndBindPSP(f *framework.Framework, pspTemplate string) (cleanup func()) {
+	// Create the PodSecurityPolicy object.
+	json, err := utilyaml.ToJSON([]byte(pspTemplate))
```
could define the object literal above instead of using yaml (will let the compiler help you), and use `DeepCopy()` to get a working copy

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,333 @@
+		}},
+	})
+	framework.ExpectNoError(err, "Failed to create PSP rolebinding")
+
```
should probably wait here for the authorization cache to observe this change. see use of `framework.WaitForAuthorizationUpdate`

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,333 @@
+	framework.ExpectNoError(err, "Failed to create PSP")
+
+	// Create the Role to bind it to the namespace.
+	_, err = f.ClientSet.RbacV1beta1().Roles(ns).Create(&rbacv1beta1.Role{
```
we have helper framework methods for binding clusterroles (BindClusterRole/BindClusterRoleInNamespace)... wonder if these should be made helpers as well

notably, those tolerate failure, since RBAC is not required to be enabled for e2e... if you're requiring it for this, we might want to fail earlier

liggitt commented on test/e2e/framework/psp_util.go:
```diff
@@ -0,0 +1,71 @@
+	By(fmt.Sprintf("Binding the %s PodSecurityPolicy to the default service account in %s",
+		PodSecurityPolicyPrivileged, namespace))
+	BindClusterRoleInNamespace(f.ClientSet.RbacV1beta1(),
+		PodSecurityPolicyPrivilegedRole,
```
is this still depending on the privileged policy loaded by the addon? would be better if the e2e was independent (created a privileged policy and clusterrole named `e2e-test-privileged-psp` if it didn't already exist, and bound to that)

tallclair reviewed (commented):
Done. I included a cherry-pick of https://github.com/kubernetes/kubernetes/pull/53034, but expect it to merge soon.

tallclair committed:
Add GCP addon PodSecurityPolicies & Bindings

tallclair committed:
GCP PodSecurityPolicy configuration

tallclair committed:
PodSecurityPolicy E2E tests

tallclair commented:
Rebased & squashed

liggitt commented on cluster/saltbase/salt/kube-addons/init.sls:
```diff
@@ -165,6 +165,28 @@ addon-dir-create:
 {% endif %}
 
+{% if pillar.get('enable_pod_security_policy', '').lower() == 'true' %}
+/etc/kubernetes/addons/podsecuritypolicy:
```
`.../addons/podsecuritypolicies`?

tallclair replied:
Actually, I have no idea where this is actually used, but changed anyway.

liggitt commented on cluster/saltbase/salt/kube-addons/init.sls:
```diff
@@ -165,6 +165,28 @@ addon-dir-create:
+{% endif %}
+
+{% if pillar.get('enable_pod_security_policy', '').lower() == 'true' %}
+/etc/kubernetes/addons/podsecuritypolicy-rbac:
```
not seeing `podsecuritypolicy-rbac` referenced elsewhere

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,315 @@
+var (
+	restrictivePSPTemplate = &extensionsv1beta1.PodSecurityPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "restrictive",
```
nit: qualify name to be e2e-specific (`e2e.restrictive`, etc)

liggitt replied:
nevermind, saw you qualified it per test

liggitt commented on test/e2e/auth/pod_security_policy.go:
```diff
@@ -0,0 +1,315 @@
+
+	It("should allow pods under the privileged PodSecurityPolicy", func() {
+		By("Creating & Binding a privileged policy for the test service account")
+		_, cleanup := createAndBindPSP(f, restrictivePSPTemplate)
```
is this supposed to be binding the privileged PSP?

liggitt replied:
nevermind again, you're binding both restrictive and permissive like I asked you to :)

tallclair replied:
added a comment.

liggitt commented on test/e2e/framework/psp_util.go:
```diff
@@ -0,0 +1,141 @@
+)
+
+const (
+	PodSecurityPolicyPrivileged = "e2e-test-privileged-psp"
```
nit: don't export this

tallclair replied:
done

liggitt commented on test/e2e/framework/psp_util.go:
```diff
@@ -0,0 +1,141 @@
+	privilegedPSPOnce sync.Once
+)
+
+func CreateDefaultPSPBinding(f *Framework, namespace string) {
```
nit, either don't export this or make the name clearer that it is creating a privileged test PSP

tallclair replied:
Done.

liggitt commented on test/e2e/framework/framework.go:
```diff
@@ -373,6 +374,11 @@ func (f *Framework) CreateNamespace(baseName string, labels map[string]string) (
 		f.namespacesToDelete = append(f.namespacesToDelete, ns)
 	}
+
+	if !f.SkipDefaultPSPBinding {
```
"Default" is unclear... make the name clearer, like `SkipPermissivePSPBinding`

tallclair replied:
done

tallclair reviewed (commented):
Done. I realized there are a bunch of optional addons that need policies that I missed, including:

- ip-masq-agent - https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/ip-masq-agent
- heapster (gke version) - https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/cluster-monitoring/google/heapster-controller.yaml#L141
- calico - https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/calico-policy-controller/calico-node-daemonset.yaml#L25

We'll need to add some additional config machinery to add the policies if both the PSP and the addon are enabled. I'll tackle this in a follow up PR.

liggitt commented on test/e2e/framework/psp_util.go:
```diff
@@ -0,0 +1,141 @@
+			Namespace: namespace,
+			Name:      "default",
+		})
+	ExpectNoError(WaitForNamedAuthorizationUpdate(f.ClientSet.AuthorizationV1beta1(),
```
this is timing out... check for an error on the BindClusterRoleInNamespace call?

tallclair replied:
I should have been checking for a notfound error above when creating the privileged policy, rather than checking for a non-nil return. Fixed.

liggitt commented:
failing with
```
test/e2e/framework/framework.go:134
Expected error:
    <*errors.errorString | 0xc4202e3430>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
not to have occurred
test/e2e/framework/psp_util.go:138
```

liggitt commented:
spurious `add apply --force --overwrite deployment lables e2e test enhance` commit as well?

liggitt commented:
LGTM once e2e test timeout is resolved

tallclair commented:
E2E's passing, spurious commit removed. Ready for that LGTM :grin:

liggitt commented:
/lgtm

liggitt commented:
thanks for all the work on this (@q-lee as well)

update the description to remove the bit about it not being on in CI (since it is), and add a release note about `kube-up` on gce support for `ENABLE_POD_SECURITY_POLICY=true`

tallclair commented:
Updated. w00t!
